File: Freedom/大众点评/首页/DZHomeController.m
//  DZHomeController.m
//  Freedom
//  Created by Super on 15/12/1.
#import "DZHomeController.h"
#import <XCategory/UILabel+expanded.h>
@interface DZHomeViewCell1 : BaseCollectionViewCell
@end
@implementation DZHomeViewCell1
-(void)initUI{//120
    self.icon = [[UIImageView alloc]initWithFrame:CGRectMake(10, 60, W(self)/2-10,60)];
    UIView *view = [self getViewWithFrame:CGRectMake(0, 0, APPW/2-10, 55)];
    UIView *view2 = [self getViewWithFrame:CGRectMake(APPW/2, 0, APPW/2-10, 55)];
    UIView *view3 = [self getViewWithFrame:CGRectMake(X(view2), YH(view2)+10, W(view2), H(view2))];
    UIImageView *image1 = [[UIImageView alloc]initWithFrame:CGRectMake(W(view2)-40, 0, 40, H(view2))];
    image1.image = [UIImage imageNamed:@"image1.jpg"];
    UIImageView *image2 = [[UIImageView alloc]initWithFrame:CGRectMake(W(view2)-40, 0, 40, H(view2))];
    image2.image = [UIImage imageNamed:@"image2.jpg"];
    [view2 addSubview:image1];
    [view3 addSubview:image2];
    self.line = [[UIView alloc]initWithFrame:CGRectMake(0, 198, APPW, 2)];
    self.line.backgroundColor = whitecolor;
    [self addSubviews:view,view2,self.icon,view3,nil];
}
-(UIView*)getViewWithFrame:(CGRect)rect{
    UIView *view = [[UIView alloc]initWithFrame:rect];
    UILabel *a = [UILabel labelWithFrame:CGRectMake(10, 0, APPW/2-20, 18) font:fontnomal color:redcolor text:@"外卖贺新春"];
    UILabel *b = [UILabel labelWithFrame:CGRectMake(X(a), YH(a), W(a), H(a)) font:fontnomal color:blacktextcolor text:@"省事省力又省心"];
    UILabel *c = [UILabel labelWithFrame:CGRectMake(X(a), YH(b), 100, 15) font:fontnomal color:yellowcolor text:@"用外卖订年夜饭"];
    c.layer.cornerRadius = 7;c.layer.borderWidth=1;c.layer.borderColor = redcolor.CGColor;
    [view addSubviews:a,b,c,nil];
    return view;
}
-(void)setCollectionDataWithDic:(NSDictionary *)dict{
    self.icon.image = [UIImage imageNamed:@"image4.jpg"];
}
@end
@interface DZHomeViewCell2 : BaseCollectionViewCell
@end
@implementation DZHomeViewCell2
-(void)initUI{//100
    self.icon = [[UIImageView alloc]initWithFrame:CGRectMake(0, 0, APPW/4,60)];
    self.title = [[UILabel alloc]initWithFrame:CGRectMake(20, YH(self.icon), W(self.icon), 20)];
    self.script = [[UILabel alloc]initWithFrame:CGRectMake(X(self.title),YH(self.title), W(self.title), H(self.title))];
    self.script.font = fontnomal;
    self.title.font = fontSmallTitle;self.script.textColor = gradtextcolor;
    [self addSubviews:self.icon,self.title,self.script,nil];
}
-(void)setCollectionDataWithDic:(NSDictionary *)dict{
    self.title.text = @"全球贺新年";
    self.script.text = @"春节专享";
    self.icon.image = [UIImage imageNamed:@"taobaomini2"];
}
@end
@interface DZHomeViewCell3 : BaseCollectionViewCell
@end
@implementation DZHomeViewCell3
-(void)initUI{//80
    self.icon = [[UIImageView alloc]initWithFrame:CGRectMake(0, 0, APPW/4-11,50)];
    self.icon.clipsToBounds = YES;self.icon.layer.cornerRadius = 10;
    self.title = [[UILabel alloc]initWithFrame:CGRectMake(0, YH(self.icon), W(self.icon), 18)];
    self.script = [[UILabel alloc]initWithFrame:CGRectMake(X(self.title),YH(self.title), W(self.title), 15)];
    self.script.font = Font(12);
    self.title.font = fontnomal;self.script.textColor = gradtextcolor;
    self.title.textAlignment = NSTextAlignmentCenter;self.script.textAlignment = NSTextAlignmentCenter;
    [self addSubviews:self.icon,self.title,self.script,nil];
}
-(void)setCollectionDataWithDic:(NSDictionary *)dict{
    self.title.text = @"爱车";
    self.script.text = @"9.9元洗车";
    self.icon.image = [UIImage imageNamed:@"taobaomini1"];
}
@end
@interface DZHomeViewCell4 : BaseCollectionViewCell
@end
@implementation DZHomeViewCell4
-(void)initUI{//100
    self.icon = [[UIImageView alloc]initWithFrame:CGRectMake(10, 0, APPW/4,80)];
    self.title = [[UILabel alloc]initWithFrame:CGRectMake(XW(self.icon)+10, Y(self.icon), APPW-XW(self.icon)-20, 20)];
    self.script = [[UILabel alloc]initWithFrame:CGRectMake(X(self.title), YH(self.title), W(self.title), 40)];
    UILabel *a = [[UILabel alloc]initWithFrame:CGRectMake(APPW-80, Y(self.title), 70, H(self.title))];
    a.textAlignment = NSTextAlignmentRight;
    UILabel *b = [[UILabel alloc]initWithFrame:CGRectMake(X(self.title), YH(self.script), W(self.title), H(self.title))];
    UILabel *d = [[UILabel alloc]initWithFrame:CGRectMake(X(a), Y(b), W(a), H(a))];
    d.textAlignment = NSTextAlignmentRight;
    self.script.font = a.font = d.font = fontnomal;
    self.script.numberOfLines =0;
    self.script.textColor = a.textColor = d.textColor = graycolor;
    b.textColor = redcolor;
    a.text = @"575m";b.text = @"￥69";
    d.text = @"已售50000";
    UIView *ling = [[UIView alloc]initWithFrame:CGRectMake(10,99, APPW-20, 1)];
    ling.backgroundColor = whitecolor;
    [self addSubviews:self.icon,self.title,a,self.script,b,d,ling,nil];
}
-(void)setCollectionDataWithDic:(NSDictionary *)dict{
    self.title.text = @"上海海洋水族馆(4A)";
    self.script.text = @"[陆家嘴]4.2分|门票、套餐、线路游 等优惠，欢迎上门体验";
    self.icon.image = [UIImage imageNamed:PuserLogo];
}
@end
@interface DZHomeHeadView1 : UICollectionReusableView{
    BaseScrollView *DZtoutiaoV;
}
@end
@implementation DZHomeHeadView1
- (id)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {[self initUI];}return self;
}
-(void)initUI{
  
    DZtoutiaoV = [BaseScrollView sharedViewBannerWithFrame:CGRectMake(20, 0, APPW-40, 60) viewsNumber:5 viewOfIndex:^UIView *(NSInteger index) {
        UIView *view = [[UIView alloc]initWithFrame:CGRectMake(0, 0, APPW-50, 60)];
        UIImageView *icon = [[UIImageView alloc]initWithFrame:CGRectMake(10, 10, 40,40)];
        icon.clipsToBounds = YES;icon.layer.cornerRadius = 20;
        icon.image = [UIImage imageNamed:PuserLogo];
        UILabel *label1 = [UILabel labelWithFrame:CGRectMake(XW(icon)+10, 10, APPW/2, 40) font:fontnomal color:blacktextcolor text:@"好友蜂蜜绿茶，吃完这家，还有下一家。地点中环广场店"];
        label1.numberOfLines = 0;
        [view addSubviews:icon,label1,nil];
        view.backgroundColor = redcolor;
        if(index%2){view.backgroundColor = yellowcolor;}
        return view;
    } Vertically:YES setFire:YES];
    [self addSubview:DZtoutiaoV];
}
@end
@interface DZHomeHeadView2 : UICollectionReusableView{
    UILabel *titleLabel;
}
@end
@implementation DZHomeHeadView2
- (id)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {[self initUI];}return self;
}
-(void)initUI{
    titleLabel = [[UILabel alloc]initWithFrame:CGRectMake(0, 5, APPW, 20)];
    titleLabel.textColor = redcolor;
    titleLabel.textAlignment = NSTextAlignmentCenter;
    titleLabel.text = @"为你优选BEST";
    self.backgroundColor = whitecolor;
    [self addSubview:titleLabel];
}
@end
//FIXME:大众点评正式VC
@interface DZHomeController ()<UICollectionViewDelegateFlowLayout>{
    BaseScrollView *itemScrollView;
}
@end
@implementation DZHomeController
- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    UIBarButtonItem *more = [[UIBarButtonItem alloc]initWithImage:[UIImage imageWithRenderingOriginalName:@"u_add_y"] style:UIBarButtonItemStylePlain actionBlick:^{}];
    self.navigationItem.rightBarButtonItem = more;
    UIBarButtonItem *map = [[UIBarButtonItem alloc]initWithTitle:@"北京" style:UIBarButtonItemStylePlain actionBlick:^{}];
    self.navigationItem.leftBarButtonItem = map;
    UISearchBar *searchBar = [[UISearchBar alloc] init];
    searchBar.placeholder = @"输入商户名、地点";
    self.navigationItem.titleView = searchBar;
    NSArray *titles = @[@"美食",@"电影",@"酒店",@"休闲娱乐",@"外卖",@"机票/火车票",@"丽人",@"周边游",@"亲子",@"KTV",@"高端酒店",@"足疗按摩",@"结婚",@"家族",@"学习培训",@"景点",@"游乐园",@"生活服务",@"洗浴",@"全部分类"];
    NSArray *icons = @[@"taobaomini1",@"taobaomini2",@"taobaomini3",@"taobaomini4",@"taobaomini5",@"taobaomini1",@"taobaomini2",@"taobaomini3",@"taobaomini4",@"taobaomini5",@"taobaomini1",@"taobaomini2",@"taobaomini3",@"taobaomini4",@"taobaomini5",@"taobaomini1",@"taobaomini2",@"taobaomini3",@"taobaomini4",@"taobaomini5"];
    itemScrollView = [BaseScrollView sharedScrollItemWithFrame:CGRectMake(0, 60, APPW, 200) icons:icons titles:titles size:CGSizeMake(APPW/5.0, 70) hang:2 round:YES];
    itemScrollView.backgroundColor = whitecolor;
    BaseCollectionViewLayout *layout = [BaseCollectionViewLayout sharedFlowlayoutWithCellSize:CGSizeMake((APPW-50)/4, 90) groupInset:UIEdgeInsetsMake(10, 10, 0, 10) itemSpace:10 linespace:10];
    self.collectionView = [[BaseCollectionView alloc]initWithFrame:CGRectMake(0, 0, APPW, APPH-110) collectionViewLayout:layout];
    self.collectionView.dataArray = [NSMutableArray arrayWithObjects:@{@"name":@"流量充值",@"pic":PuserLogo}, nil];
    [self fillTheCollectionViewDataWithCanMove:NO sectionN:4 itemN:20 itemName:@"DZHomeViewCell1"];
    [self.collectionView registerClass:[DZHomeViewCell2 class] forCellWithReuseIdentifier:@"DZHomeViewCell2"];
    [self.collectionView registerClass:[DZHomeViewCell3 class] forCellWithReuseIdentifier:@"DZHomeViewCell3"];
    [self.collectionView registerClass:[DZHomeViewCell4 class] forCellWithReuseIdentifier:@"DZHomeViewCell4"];
    [self.collectionView registerClass:[DZHomeHeadView1 class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"headview1"];
    [self.collectionView registerClass:[DZHomeHeadView2 class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"headview2"];
    [self.collectionView addSubview:itemScrollView];
    self.collectionView.dataSource = self;
    self.collectionView.delegate = self;
    [self.view addSubview:self.collectionView];
}
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    if (section == 0)return 1;if (section == 1)return 6;if (section == 2)return 8;return 20;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    BaseCollectionViewCell *cell = nil;
    if(indexPath.section == 0){
        cell =  [collectionView dequeueReusableCellWithReuseIdentifier:self.collectionReuseId forIndexPath:indexPath];
        [cell  setCollectionDataWithDic:nil];
    }else if (indexPath.section == 1) {
        cell =  [collectionView dequeueReusableCellWithReuseIdentifier:@"DZHomeViewCell2" forIndexPath:indexPath];
        [cell  setCollectionDataWithDic:nil];
    }else if (indexPath.section == 2) {
        cell =  [collectionView dequeueReusableCellWithReuseIdentifier:@"DZHomeViewCell3" forIndexPath:indexPath];
        [cell  setCollectionDataWithDic:nil];
    }else{
        cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"DZHomeViewCell4" forIndexPath:indexPath];
        [cell setCollectionDataWithDic:nil];
    }
    return cell;
}
//FIXME: collectionViewDelegate
//item 尺寸
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 0) {
        return CGSizeMake(APPW, 130);
    }else if(indexPath.section==1){
        return CGSizeMake(APPW/3-15, 100);
    }else if(indexPath.section==2){
        return CGSizeMake(APPW/4-15, 90);
    }else{
        return CGSizeMake(APPW, 100);
    }
}
//分区头的尺寸
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section {
    if (section == 0) {return CGSizeMake(APPW, 60);
    }else if(section==1){
        return CGSizeMake(APPW, 30);
    }else if(section==2){
        return CGSizeMake(APPW, 60);
    }else if(section==3){
        return CGSizeMake(APPW, 30);
    }
    return CGSizeMake(APPW,30);
}
//分区尾的尺寸
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout referenceSizeForFooterInSection:(NSInteger)section {
    return CGSizeZero;
}
-(UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section{
    if(section==0){
        return UIEdgeInsetsMake(160, 10, 0, 10);
    }else{
        return UIEdgeInsetsMake(10, 10, 0, 10);
    }
}
- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath{
    if (kind == UICollectionElementKindSectionHeader){
        if(indexPath.section==0){
            return [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"headview1" forIndexPath:indexPath];
        }else if(indexPath.section==1){
            return [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"headview2" forIndexPath:indexPath];
        }else if(indexPath.section==2){
            return [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"headview1" forIndexPath:indexPath];
        }else{
            return [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"headview2" forIndexPath:indexPath];
    }
    }
    return nil;
}
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    NSString *log = [NSString stringWithFormat:@"你选择的是%zd，%zd", indexPath.section, indexPath.row];
    [SVProgressHUD showSuccessWithStatus:log];DLog(@"%@",log);
}
@end

File: Freedom/大众点评/首页/DZHomeController.swift
//
//  DZHomeController.swift
//  Freedom
import UIKit
//import XExtension
class DZHomeViewCell1:BaseCollectionViewCell<Any> {
    override func initUI() {//120
        icon = UIImageView(frame: CGRect(x: 10, y: 60, width: self.width / 2 - 10, height: 60))
        let view = getViewWithFrame(CGRect(x: 0, y: 0, width: APPW / 2 - 10, height: 55))
        let view2 = getViewWithFrame(CGRect(x: APPW / 2, y: 0, width: APPW / 2 - 10, height: 55))
        let view3 = getViewWithFrame(CGRect(x: view2.x, y: view2.bottom + 10, width: view2.width, height: view2.height))
        let image1 = UIImageView(frame: CGRect(x: view2.width - 40, y: 0, width: 40, height: view2.height))
        image1.image = Image.logo.image
        let image2 = UIImageView(frame: CGRect(x: view2.width - 40, y: 0, width: 40, height: view2.height))
        image2.image = Image.logo.image
        view2.addSubview(image1)
        view3.addSubview(image2)
        addSubviews([view, view2, icon, view3])
    }
    func getViewWithFrame(_ rect: CGRect) -> UIView {
        let view = UIView(frame: rect)
        let a = UILabel(frame: CGRect(x: 10, y: 0, width: APPW / 2 - 20, height: 18), font: .small, color: .red, text: "外卖贺新春")
        let b = UILabel(frame: CGRect(x: a.x, y: a.bottom, width: a.width, height: a.height), font: .small, color: .dark, text: "省事省力又省心")
        let c = UILabel(frame: CGRect(x: a.x, y: b.bottom, width: 100, height: 15), font: .small, color: .yellow, text: "用外卖订年夜饭")
        c.layer.cornerRadius = 7
        c.layer.borderWidth = 1
        c.layer.borderColor = UIColor.redx.cgColor
        view.addSubviews([a, b, c])
        return view
    }
    func setCollectionDataWithDic(_ dict: [AnyHashable: Any]?) {
        icon.image = TBImage.im4.image
    }
}
class DZHomeViewCell2:BaseCollectionViewCell<Any> {
    override func initUI() {//100
        icon = UIImageView(frame: CGRect(x: 0, y: 0, width: APPW / 4, height: 60))
        title = UILabel(frame: CGRect(x: 20, y: icon.bottom, width: icon.width, height: 20))
        script = UILabel(frame: CGRect(x: title.x, y: title.bottom, width: title.width, height: title.height))
        script.font = .small
        title.font = .middle
        script.textColor = .cd
        addSubviews([icon, title, script])
    }
    func setCollectionDataWithDic(_ dict: [AnyHashable: Any]?) {
        title.text = "全球贺新年"
        script.text = "春节专享"
        icon.image = Image.logo.image
    }
}
class DZHomeViewCell3:BaseCollectionViewCell<Any> {
    override func initUI() {//80
        icon = UIImageView(frame: CGRect(x: 0, y: 0, width: APPW / 4 - 11, height: 50))
        icon.clipsToBounds = true
        icon.layer.cornerRadius = 10
        title = UILabel(frame: CGRect(x: 0, y: icon.bottom, width: icon.width, height: 18))
        script = UILabel(frame: CGRect(x: title.x, y: title.bottom, width: title.width, height: 15))
        script.font = Font(12)
        title.font = .small
        script.textColor = .cd
        title.textAlignment = .center
        script.textAlignment = .center
        addSubviews([icon, title, script])
    }
    func setCollectionDataWithDic(_ dict: [AnyHashable: Any]?) {
        title.text = "爱车"
        script.text = "9.9元洗车"
        icon.image = Image.logo.image
    }
}
class DZHomeViewCell4:BaseCollectionViewCell<Any> {
    override func initUI() {//100
        icon = UIImageView(frame: CGRect(x: 10, y: 0, width: APPW / 4, height: 80))
        title = UILabel(frame: CGRect(x: icon.right + 10, y: icon.y, width: APPW - icon.right - 20, height: 20))
        script = UILabel(frame: CGRect(x: title.x, y: title.bottom, width: title.width, height: 40))
        let a = UILabel(frame: CGRect(x: APPW - 80, y: title.y, width: 70, height: title.height))
        a.textAlignment = .right
        let b = UILabel(frame: CGRect(x: title.x, y: script.bottom, width: title.width, height: title.height))
        let d = UILabel(frame: CGRect(x: a.x, y: b.y, width: a.width, height: a.height))
        d.textAlignment = .right
        d.font = .small
        a.font = d.font
        script.font = a.font
        script.numberOfLines = 0
        d.textColor = .gray
        a.textColor = d.textColor
        script.textColor = a.textColor
        b.textColor = .red
        a.text = "575m"
        b.text = "￥69"
        d.text = "已售50000"
        let ling = UIView(frame: CGRect(x: 10, y: 99, width: APPW - 20, height: 1))
        ling.backgroundColor = .white
        addSubviews([icon, title, a, script, b, d, ling])
    }
    func setCollectionDataWithDic(_ dict: [AnyHashable: Any]?) {
        title.text = "上海海洋水族馆(4A)"
        script.text = "[陆家嘴]4.2分|门票、套餐、线路游 等优惠，欢迎上门体验"
        icon.image = Image.logo.image
    }
}
class DZHomeHeadView1: UICollectionReusableView {
    var DZtoutiaoV: BaseScrollView?
    override init(frame: CGRect) {
        super.init(frame: frame)
        initUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    func initUI() {
        DZtoutiaoV = BaseScrollView(viewBanner: CGRect(x: 20, y: 0, width: APPW - 40, height: 60), viewsNumber: 5, viewOfIndex: {(_ index: Int) -> UIView in
            let view = UIView(frame: CGRect(x: 0, y: 0, width: APPW - 50, height: 60))
            let icon = UIImageView(frame: CGRect(x: 10, y: 10, width: 40, height: 40))
            icon.clipsToBounds = true
            icon.layer.cornerRadius = 20
            icon.image = Image.logo.image
            let label1 = UILabel(frame: CGRect(x: icon.right + 10, y: 10, width: APPW / 2, height: 40), font: .small, color: .c3, text: "好友蜂蜜绿茶，吃完这家，还有下一家。地点中环广场店")
            label1.numberOfLines = 0
            view.addSubviews([icon, label1])
            view.backgroundColor = .red
            if index % 2 != 0 {
                view.backgroundColor = .yellow
            }
            return view
        }, vertically: true, setFire: true)
        addSubview(DZtoutiaoV!)
    }
}
class DZHomeHeadView2: UICollectionReusableView {
    var titleLabel: UILabel?
    override init(frame: CGRect) {
        super.init(frame: frame)
        initUI()
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    func initUI() {
        titleLabel = UILabel(frame: CGRect(x: 0, y: 5, width: APPW, height: 20))
        titleLabel?.textColor = .red
        titleLabel?.textAlignment = .center
        titleLabel?.text = "为你优选BEST"
        backgroundColor = .white
        if let aLabel = titleLabel {
            addSubview(aLabel)
        }
    }
}
final class DZHomeController: DZBaseViewController,UICollectionViewDataSource,UICollectionViewDelegate {
    var collectionView: BaseCollectionView!
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.whitex
        let more = UIBarButtonItem(image: Image.u_Add.image, style: .done, target: nil, action: nil)
        navigationItem.rightBarButtonItem = more
        let map = UIBarButtonItem(title: "北京") {
            
        }
        navigationItem.leftBarButtonItem = map
        let searchBar = UISearchBar()
        searchBar.placeholder = "输入商户名、地点"
        navigationItem.titleView = searchBar
        let titles = ["美食", "电影", "酒店", "休闲娱乐", "外卖", "机票/火车票", "丽人", "周边游", "亲子", "KTV", "高端酒店", "足疗按摩", "结婚", "家族", "学习培训", "景点", "游乐园", "生活服务", "洗浴", "全部分类"]
        let icons = ["taobaomini1", "taobaomini2", "taobaomini3", "taobaomini4", "taobaomini5", "taobaomini1", "taobaomini2", "taobaomini3", "taobaomini4", "taobaomini5", "taobaomini1", "taobaomini2", "taobaomini3", "taobaomini4", "taobaomini5", "taobaomini1", "taobaomini2", "taobaomini3", "taobaomini4", "taobaomini5"]
        let itemScrollView = BaseScrollView(scrollItem: CGRect(x: 0, y: 60, width: APPW, height: 200), icons: icons, titles: titles, size: CGSize(width: APPW / 5.0, height: 70), hang: 2, round: true)
        itemScrollView.backgroundColor = .white
        let layout = UICollectionViewFlowLayout()
        layout.itemSize = CGSize(width: (APPW - 50) / 4, height: 90)
        layout.sectionInset = UIEdgeInsets(top: 10, left: 10, bottom: 0, right: 10)
        layout.minimumInteritemSpacing = 10
        layout.minimumLineSpacing = 10
        collectionView = BaseCollectionView(frame: CGRect(x: 0, y: 0, width: APPW, height: APPH - 110), collectionViewLayout: layout)
        collectionView?.register(DZHomeViewCell1.self, forCellWithReuseIdentifier: DZHomeViewCell1.identifier)
        collectionView?.register(DZHomeViewCell2.self, forCellWithReuseIdentifier: DZHomeViewCell2.identifier)
        collectionView?.register(DZHomeViewCell3.self, forCellWithReuseIdentifier: DZHomeViewCell3.identifier)
        collectionView?.register(DZHomeViewCell4.self, forCellWithReuseIdentifier: DZHomeViewCell4.identifier)
        collectionView?.register(DZHomeHeadView1.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: DZHomeHeadView1.identifier)
        collectionView?.register(DZHomeHeadView2.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: DZHomeHeadView2.identifier)
        collectionView?.addSubview(itemScrollView)
        collectionView?.dataSource = self
        collectionView?.delegate = self
        if let aView = collectionView {
            view.addSubview(aView)
        }
    }
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if section == 0 {
            return 1
        }
        if section == 1 {
            return 6
        }
        if section == 2 {
            return 8
        }
        return 20
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        var cell: BaseCollectionViewCell<Any>? = nil
        if indexPath.section == 0 {
            cell = collectionView.dequeueCell(DZHomeViewCell1.self, for: indexPath)
        } else if indexPath.section == 1 {
            cell = collectionView.dequeueCell(DZHomeViewCell2.self, for: indexPath)
        } else if indexPath.section == 2 {
            cell = collectionView.dequeueCell(DZHomeViewCell3.self, for: indexPath)
        } else {
            cell = collectionView.dequeueCell(DZHomeViewCell4.self, for: indexPath)
        }
        return cell!
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        if indexPath.section == 0 {
            return CGSize(width: APPW, height: 130)
        } else if indexPath.section == 1 {
            return CGSize(width: APPW / 3 - 15, height: 100)
        } else if indexPath.section == 2 {
            return CGSize(width: APPW / 4 - 15, height: 90)
        } else {
            return CGSize(width: APPW, height: 100)
        }
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        if section == 0 {
            return CGSize(width: APPW, height: 60)
        } else if section == 1 {
            return CGSize(width: APPW, height: 30)
        } else if section == 2 {
            return CGSize(width: APPW, height: 60)
        } else if section == 3 {
            return CGSize(width: APPW, height: 30)
        }
        return CGSize(width: APPW, height: 30)
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForFooterInSection section: Int) -> CGSize {
        return CGSize.zero
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        if section == 0 {
            return UIEdgeInsets(top: 160, left: 10, bottom: 0, right: 10)
        } else {
            return UIEdgeInsets(top: 10, left: 10, bottom: 0, right: 10)
        }
    }
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        if kind == UICollectionView.elementKindSectionHeader {
            if indexPath.section == 0 {
                return collectionView.dequeueHeadFoot(Headview1.self, kind: kind, for: indexPath)
            } else if indexPath.section == 1 {
                return collectionView.dequeueHeadFoot(Headview2.self, kind: kind, for: indexPath)
            } else if indexPath.section == 2 {
                return collectionView.dequeueHeadFoot(Headview1.self, kind: kind, for: indexPath)
            } else {
                return collectionView.dequeueHeadFoot(Headview2.self, kind: kind, for: indexPath)
            }
        }else{
            return collectionView.dequeueHeadFoot(Headview1.self, kind: kind, for: indexPath)
        }
    }
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let log = "你选择的是\(indexPath.section)，\(indexPath.row)"
        Dlog(log)
    }
}
